### ISSUE TITLE
Provide a get_number() functionality to On Demand

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#1703]
Currently, the On Demand API tells you that a token is a number, if you ask for the type, and then it allows to parse it as a float or an integer.

However, this makes it hard for a user to deal with inputs such as `[1,1.]` if they want to detect that they have an integer followed by a float.

So how do we solve this? I think we should have some kind of `ondemand::number` type which can contain either an integer or a float.

So you would do...

```C++
ondemand::number number = value.get_number();
if(number.is_uint64()) {
  uint64_t x = number.get_uint64();
} else if (number.is_int64()) {
  int64_t x = number.get_int64();
} else {
  double d = number.get_double();
}
``` 

Thus you could use the On Demand API as you use the DOM API.

I think it is damn ugly, but I cannot think of a simpler way.

Underneath, the number type would be a union with a type indicator. I can implement this (complete with the tests) in a few hours. So it is not the implementation that is an issue... we all have the code... it is the API that I am not fond on.

Feedback invited.

cc @eddelbuettel @NicolasJiaxin @jkeiser

**Comment by lemire** [comment]
I think we need to resolve this before 1.0.

**Comment by eddelbuettel** [comment]
I had a similar puzzle recently in parsing R expression for selections and aiming to infer what type it may be.  An example would be  `a > 2` and guestimating if a may be an `int` or `double` variable.  I concluded one cannot reliably without having type information.  

In your example of `[1, 1.]` maybe users want two floats but just wrote "in short-hand" ?  

With JSON not being typed ... this is tricky.

**Comment by NicolasJiaxin** [comment]
I also thought about doing it this way, which in the worst case means that we are parsing numbers twice. I do not know if this is feasible, but would it be possible to parse numbers only once on "dispatch". By that, I mean we scan a number assuming it is an integer, and if we ever encounter a decimal point or something like that, continue the parsing, but consider now the first part as a the integral part of said double. And if there is no decimal point, just continue parsing normally and return as a integer. For example, if we have something like `324.54`, we would first parse `324` as a unsigned integer. Then, when we get to the `.`, we can now consider the unsigned integer `324` as the first part and continue parsing to get the second part of the double.  We would also check if the first digit `'-'` or not to see if we need uint64 or int64. However, like I said, I do not know if this is feasible in C++ code.

**Comment by lemire** [comment]
@NicolasJiaxin Thanks for the feedback.

We do not want to parse twice. I am puking in a bucket right now at the thought of parsing data twice. (joke)

This is actually super easy for me to implement and in my proposal, we would parse just once.

The way we do it in DOM is this:

1. If has no decimal separator or exponent, then we treat it as an integer (int64_t).
2. If it is an integer, but it overflows and is positive, try to represent it as an uint64_t.
3. If it has a decimal separator or an exponent, go to double.

One thing I might do differently from DOM, if I use the same approach is the "-0" case. I think that this indicates a double (because there is no negative 0 in integers).

So that's what I might implement for On Demand.

It is not hard for me. I expect I can do it in hours. What is a bit harder is to make sure it has good usability.

(My current plan is to start the implementation on Monday. I wanted to create the issue first so that people can comment on the design issues.)

**Comment by NicolasJiaxin** [comment]
Another proposal that I have is that we could have something like `get_number_type()` instead of `get_number()` which would just tell the user the type of number without parsing it. It would return `uint`, `int` or `double`. We could add a flag that would identify  `'-'`, `'.'` or `'e'` inside numbers at parsing (a little bit like a structural index). Then, when `get_number_type()` is called it would return something just check what conditions are true/false for a given number to determine if  `uint`, `int` or `double`. For example, `324.5` has `.` true so it must be a double. `-23` has `-` true but `.` and `e` false so it is a signed integer. I do not think it is feasible with current API though, because adding `'.'` or `'e'` as structural indexes would probably break the code, but this is just to illustrate another concept that might be used if someone can figure out the correct way to implement it if possible.

**Comment by jkeiser** [comment]
The alternative is to have is_float() and is_negative() methods and let the user do that.

The discriminated union is a good idea in general though, for certain cases: you could even expand it to all possible types and have a get_native() or something ...

**Comment by lemire** [comment]
I like get_number_type but the issue is that to check the type you need to scan the whole number string… then you’d need to scan it all a second time to parse. 

I want a one-pass design or, at least, not lock up the API in a model where you must do at least two passes on each number string.

The get_type approach in the general API works because we do not need to parse all of an array to know it is an array… just looking at the first character is enough.

But if I have…

1468957985666

then I need to read each digit in the worst case to be sure it is not a float because there might be a decimal separator hiding in there.

Mow, of course, I could massively optimize this determination… but in the worst case, I still need to look at all characters.

So I want to parse and classify in a single pass.

**Comment by lemire** [comment]
I would still want to keep get_number_type no matter what!

**Comment by jkeiser** [comment]
Thinking about it more, get_native is really bad, as it would eliminate many optimizations the compiler can do if it knows the type of your value. This could be an argument for is_float and is_negative, too, but the argument against is_float is that it duplicates the time finding the dot, and get_number can just keep going.

**Comment by lemire** [comment]
@jkeiser @NicolasJiaxin Right. So it is really tempting to just say f*ck it and have an is_float function that is non-parsing. It may even be efficient. I am just concerned about locking us up in one direction.



I am not sure the the non-parsing type-determination function provides a better user experience: you still have two steps… you must query the type and then recover the value.

And then both steps can throw an error… You can get an error with is_float and the an error with get_double.

With the number approach above, we can make several methods error-free. After all, once you have parsed the number, you know its type for sure.

I think that the union approach could autoconvert like the DOM API does…

**Comment by NicolasJiaxin** [comment]
Yes, I think you are right about this, we should combine parsing and type determination in a single function for `get_number()`. On the other hand, to relate this issue to the RcppSimdJson wrapper, a method like `get_number_type()` would also be useful within simdjson to determine things such as if I have an array of floats or integers. We could support these two functions and be clear about the context for using them.

**Comment by lemire** [comment]
@NicolasJiaxin You make a good argument.

**Comment by lemire** [comment]
@NicolasJiaxin ETA for the first working draft of my implementation is tomorrow afternoon. It is only a few hours, but I have, unfortunately, other responsibilities. I realize that you may depend on this implementation for your work so I will not delay.
